File: leetcode/easy/String/cpp/680_Valid Palindrome II.cpp
class Solution {
public:
    bool isPalindrome(string s){
        
        int l = 0;
        int r = s.length()-1;
        while(l < r){
            if(s[l] != s[r]){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
    
    string deleteIndex(int index, string s, int n){
        
        string tmp = "";
        for(int i = 0; i < index; i++){
            tmp += s[i];
        }
        for(int i = index+1; i < n; i++){
            tmp += s[i];
        }
        
        return tmp;
    }
    
    bool validPalindrome(string s) {
        
        int l = 0;
        int r = s.length()-1;
        int n = s.length();
        while(l < r){
            if(s[l] != s[r]){
                
                string newstring = deleteIndex(l, s, n);
                
                if(isPalindrome(newstring)){
                    return true;
                }
                
                newstring = deleteIndex(r, s, n);
                
                return isPalindrome(newstring);
            }
            l++;
            r--;
        }
        
        return true;
    }
};

File: leetcode/easy/greedy/680_Valid Palindrome II.java
class Solution {
    public boolean validPalindrome(String s) {

        int n = s.length();
        int l = 0, r = n-1;
        while(l < r){

            if(s.charAt(l) != s.charAt(r)){

                String newString = deleteChar(s, l);

                if(isPalindrome(newString)) return true;

                newString = deleteChar(s, r);

                return (isPalindrome(newString));
            }

            l++;
            r--;
        }

        return true;
    }

    private boolean isPalindrome(String newString) {

        int l = 0;
        int r = newString.length()-1;
        while(l < r){
            if(newString.charAt(l) != newString.charAt(r)){
                return false;
            }
            l++;
            r--;
        }
        return true;

    }

    private String deleteChar(String s, int deleteIdx) {

        String newString = "";
        int n = s.length();
        newString += s.substring(0, deleteIdx);
        newString += s.substring(deleteIdx+1, n);

        return newString;
    }
}
